nxp: usb: zephyr support on imx93 frdm Cortex-A Core

JonyZhang7 opened the pull request:
nxp: add usb support on frdm-imx93

Depends on https://github.com/zephyrproject-rtos/hal_nxp/pull/615 (has been merged)

JasonHe-nxp committed:
drivers: clock: mcux_ccm: add USB clock control support

Add USB_CLK and USB_PHY_CLK support for NXP EHCI controller on MIMX9352.

Signed-off-by: Jason He <jason.he_1@nxp.com>
Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>
Signed-off-by: Jony Zhang <jony.zhang@nxp.com>

JasonHe-nxp committed:
drivers: usb: mcux_ehci: add clock control for i.MX 93 family

Support device tree specified clock rates for USB controller and PHY.

Signed-off-by: Jason He <jason.he_1@nxp.com>
Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>
Signed-off-by: Jony Zhang <jony.zhang@nxp.com>

JasonHe-nxp committed:
boards: frdm_imx93: enable USB support

Add EHCI controller configuration with dual USB instances (USB1/USB2),
device tree bindings, 8 bidirectional endpoints, and UDC workqueue
stack optimization for frdm_imx93 A55 platform.

Signed-off-by: Jason He <jason.he_1@nxp.com>
Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>
Signed-off-by: Jony Zhang <jony.zhang@nxp.com>

JiafeiPan committed:
modules: hal_nxp: fix camke for sdk-ng middleware

As middleware.cmake could be included by mcux-sdk's CMakeLists.txt,
so use CMAKE_CURRENT_LIST_DIR to make sure the directory is correct.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>
Signed-off-by: Jony Zhang <jony.zhang@nxp.com>

mmahadevan108 commented:
@JonyZhang7 , please take a look at the CI failures.

MarkWangChinese commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -34,19 +36,30 @@ LOG_MODULE_REGISTER(udc_mcux, CONFIG_UDC_DRIVER_LOG_LEVEL);
+	clock_control_subsys_rate_t usb_clock_rate;
+	const struct device *usbphy_clock_dev;
+	clock_control_subsys_t usbphy_clock_subsys;
+	clock_control_subsys_rate_t usbphy_clock_rate;
```
For other NXP mcu platforms, it seems that the clock is initialized in the `soc.c`, could you put the clock related codes there?

JonyZhang7 commented on subsys/usb/device_next/usbd_ch9.c:
```diff
@@ -590,7 +590,7 @@ static void string_ascii7_to_utf16le(struct usbd_desc_node *const dn,
 	}
 
-	LOG_DBG("wLength %u, bLength %u, tailroom %u",
+	LOG_DBG("wLength %u, bLength %u, tailroom %zu",
```
> I think this commit should go in its own PR. Can you also provide better PR title so it is immediately clear that this adds support for new target?

Hi @tmon-nordic sorry for late, I modified PR title. tomorrow is feature freeze, there is no time left for me to submit a new PR for this patch. terribly sorry. Could you help review again, thanks.

JonyZhang7 replied:
> @JonyZhang7 Please move this commit to own PR.

Hi @jfischer-no add to PR: https://github.com/zephyrproject-rtos/zephyr/pull/98245

JonyZhang7 commented:
> @JonyZhang7 , please take a look at the CI failures.

Hi @mmahadevan108 , I fixed CI failures, thanks a lot!

jfischer-no commented on boards/nxp/frdm_imx93/frdm_imx93_mimx9352_a55.yaml:
```diff
@@ -21,6 +21,7 @@ supported:
   - can
   - net
   - watchdog
+  - usb_device
```
Replace it with `- usbd`.

JonyZhang7 commented on boards/nxp/frdm_imx93/frdm_imx93_mimx9352_a55_defconfig:
```diff
@@ -28,3 +28,6 @@ CONFIG_CONSOLE=y
 CONFIG_CLOCK_CONTROL=y
+
+# USB Except
+CONFIG_UDC_WORKQUEUE=n
```
> Why?

@JiafeiPan  do we need close udc workqueue?

JonyZhang7 replied:
> Why?

Because the work queue size was small, the stack size was changed to 4096.

JiafeiPan commented on boards/nxp/frdm_imx93/frdm_imx93_mimx9352_a55_defconfig:
```diff
@@ -28,3 +28,6 @@ CONFIG_CONSOLE=y
 CONFIG_CLOCK_CONTROL=y
+
+# USB Except
+CONFIG_UDC_WORKQUEUE_STACK_SIZE=4096
```
Define it in Kconfig.defconfig

```
configdefault UDC_WORKQUEUE_STACK_SIZE
    default 4096
```

JonyZhang7 replied:
> Define it in Kconfig.defconfig
> 
> ```
> configdefault UDC_WORKQUEUE_STACK_SIZE
>     default 4096
> ```

thanks

jfischer-no reviewed (changes requested):
> Hi, @jfischer-no , please revisit this PR, I believe your changes request have been addressed, thanks.

Not addressed.

JonyZhang7 committed:
drivers: usb: mcux_ehci: add MMIO configuration for i.MX 93 family

The original driver is only feasible for M-Core, but for Cortex-A Core,
it need to do MMU mapping to map physical address to virtual address,
to the main update is to add MMIO mapping in this driver, and all the
register access should use virtual address.

Replace direct base address access with DEVICE_MMIO_NAMED_* macros
to provide better memory mapping abstraction and improve platform
portability.

This change:
- Adds DEVICE_MMIO_NAMED_ROM/RAM to config and data structures
- Maps the register base address during driver pre-initialization
- Updates all base address references to use DEVICE_MMIO_NAMED_GET
- Maintains backward compatibility with existing functionality

The DEVICE_MMIO API provides a standardized way to handle memory-mapped
registers across different platforms and memory protection schemes.

Signed-off-by: Jason He <jason.he_1@nxp.com>
Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>
Signed-off-by: Jony Zhang <jony.zhang@nxp.com>

jfischer-no commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -859,6 +882,42 @@ static const usb_device_controller_interface_struct_t udc_mcux_if = {
+#define UDC_MCUX_CLK1_DEFINE_NULL							\
+	.phy_clock_dev = NULL,								\
+	.phy_clock_subsys = NULL,							\
+	.phy_clock_rate = NULL,
```
This macro is redundant and can be removed.

JonyZhang7 replied:
yes

igalloway commented on drivers/usb/udc/udc_mcux_ehci.c:
```diff
@@ -1,19 +1,21 @@
 /*
- * Copyright 2024 NXP
+ * Copyright 2024-2025 NXP
```
I read that the zephyr docs do not require a change, but you can add 
"If you authored substantial, original content, you may add an additional line for yourself or your organization."

Also our company has requirements that we need to add this according to our legal dept.
WE are given guidelines where we have to add notices if we touch any files:

Example:
*Copyright 2020 originalpersonorcompany
*Copyright 2021 NXP
or
*Copyright 2021-2022 NXP
or 
*Copyright 2021, 2023 NXP

jfischer-no replied:
> Yes, the change is fine :)

@kartben Should we then remove `SPDX-FileCopyrightText:` from contribution guidelines and the optional compliance check?
